planetiler-core/src/test/java/com/onthegomap/planetiler/pmtiles/PmtilesTest.java

```diff
@@ -96,44 +113,178 @@ void testRoundtripHeader() {
+
+    in.finish(config);
+    var reader = new ReadablePmtiles(bytes);
+    var header = reader.getHeader();
+    assertEquals(3, header.numAddressedTiles());
+    assertEquals(1, header.numTileContents());
+    assertEquals(2, header.numTileEntries()); // z0 and z1 are contiguous
+    assertArrayEquals(new byte[]{0xa, 0x2}, reader.getTile(0, 0, 0));
+    assertArrayEquals(new byte[]{0xa, 0x2}, reader.getTile(0, 0, 1));
+    assertArrayEquals(new byte[]{0xa, 0x2}, reader.getTile(0, 0, 2));
+
+    Set<TileCoord> coordset = reader.getAllTileCoords().stream().collect(Collectors.toSet());
+    assertEquals(3, coordset.size());
+  }
+
+  @Test
+  void testWritePmtilesUnclustered() throws IOException {
+    var bytes = new SeekableInMemoryByteChannel(0);
+    var in = WriteablePmtiles.newWriteToMemory(bytes);
+
+    var config = PlanetilerConfig.defaults();
+    in.initialize(config, new TileArchiveMetadata(new Profile.NullProfile()), new LayerStats());
+    var writer = in.newTileWriter();
+    writer.write(new TileEncodingResult(TileCoord.ofXYZ(0, 0, 1), new byte[]{0xa, 0x2}, OptionalLong.of(42)));
+    writer.write(new TileEncodingResult(TileCoord.ofXYZ(0, 0, 0), new byte[]{0xa, 0x2}, OptionalLong.of(42)));
+
+    in.finish(config);
+    var reader = new ReadablePmtiles(bytes);
+    var header = reader.getHeader();
+    assertEquals(2, header.numAddressedTiles());
+    assertEquals(1, header.numTileContents());
+    assertEquals(2, header.numTileEntries());
+    assertFalse(header.clustered());
+    assertArrayEquals(new byte[]{0xa, 0x2}, reader.getTile(0, 0, 0));
+    assertArrayEquals(new byte[]{0xa, 0x2}, reader.getTile(0, 0, 1));
+
+    Set<TileCoord> coordset = reader.getAllTileCoords().stream().collect(Collectors.toSet());
+    assertEquals(2, coordset.size());
+  }
+
+  @Test
+  void testWritePmtilesLeafDirectories() throws IOException {
+    var bytes = new SeekableInMemoryByteChannel(0);
+    var in = WriteablePmtiles.newWriteToMemory(bytes);
+
+    var config = PlanetilerConfig.defaults();
+    in.initialize(config, new TileArchiveMetadata(new Profile.NullProfile()), new LayerStats());
+    var writer = in.newTileWriter();
+
+    int ENTRIES = 20000;
+
+    for (int i = 0; i < ENTRIES; i++) {
+      writer.write(new TileEncodingResult(TileCoord.hilbertDecode(i), ByteBuffer.allocate(4).putInt(i).array(),
+        OptionalLong.empty()));
+    }
+
+    in.finish(config);
+    var reader = new ReadablePmtiles(bytes);
+    var header = reader.getHeader();
+    assertEquals(ENTRIES, header.numAddressedTiles());
+    assertEquals(ENTRIES, header.numTileContents());
+    assertEquals(ENTRIES, header.numTileEntries());
+    assert (header.leafDirectoriesLength() > 0);
+
+    for (int i = 0; i < ENTRIES; i++) {
+      var coord = TileCoord.hilbertDecode(i);
+      assertArrayEquals(ByteBuffer.allocate(4).putInt(i).array(), reader.getTile(coord.x(), coord.y(), coord.z()));
```

Comment:
might want to add a message as the third argument to this with `coord.toString()` so if this fails we'll be able to tell what tile caused the issue.